src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperCompletionService.cs

```diff
@@ -316,5 +318,77 @@ void UpdateCompletions(string tagName, TagHelperDescriptor possibleDescriptor)
                 }
             }
         }
+
+        private static IReadOnlyList<TagHelperDescriptor> FilterFullyQualifiedCompletions(IReadOnlyList<TagHelperDescriptor> possibleChildDescriptors)
+        {
+            // Iterate once through the list to tease apart fully qualified and short name TagHelpers
+            var fullyQualifiedTagHelpers = new List<TagHelperDescriptor>();
+            var shortNameTagHelpers = new HashSet<TagHelperDescriptor>(ShortNameToFullyQualifiedComparer.Instance);
+            for (var i = 0; i < possibleChildDescriptors.Count; i++)
+            {
+                var descriptor = possibleChildDescriptors[i];
+
+                if (descriptor.IsComponentFullyQualifiedNameMatch())
+                {
+                    fullyQualifiedTagHelpers.Add(descriptor);
+                }
+                else
+                {
+                    shortNameTagHelpers.Add(descriptor);
+                }
+            }
+
+            // Re-combine the short named & fully qualified TagHelpers but filter out any fully qualified TagHelpers that have a short
+            // named representation already.
+            var filteredList = new List<TagHelperDescriptor>(shortNameTagHelpers);
+            for (var i = 0; i < fullyQualifiedTagHelpers.Count; i++)
+            {
+                var fullyQualifiedTagHelper = fullyQualifiedTagHelpers[i];
+
+                if (!shortNameTagHelpers.Contains(fullyQualifiedTagHelper))
+                {
+                    // Unimported completion item that isn't represented in a short named form.
+                    filteredList.Add(fullyQualifiedTagHelper);
+                }
+                else
+                {
+                    // There's already a shortname variant of this item, don't include it.
+                }
+            }
+
+            return filteredList;
+        }
+
+        private class ShortNameToFullyQualifiedComparer : IEqualityComparer<TagHelperDescriptor>
+        {
+            public static readonly ShortNameToFullyQualifiedComparer Instance = new ShortNameToFullyQualifiedComparer();
+
+            public bool Equals(TagHelperDescriptor x, TagHelperDescriptor y)
+            {
+                if (x == y)
```

Comment:
Shouldn't this be `x.Equals(y)` or `Object.ReferenceEquals(x, y)` depending on your intent?